src/backend/src/main/kotlin/org/icpclive/service/analytics/AnalyticsGenerator.kt

```diff
@@ -0,0 +1,150 @@
+package org.icpclive.service.analytics
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.decodeFromStream
+import org.icpclive.api.*
+import org.icpclive.util.getLogger
+import java.nio.file.Path
+import kotlin.io.path.inputStream
+
+class AnalyticsGenerator(jsonTemplatePath: Path) {
+    private val messagesTemplates = Json.decodeFromStream<JsonAnalyticTemplates>(jsonTemplatePath.inputStream())
+
+    suspend fun run(
+        analyticsMessagesFlow: FlowCollector<AnalyticsMessage>,
+        contestInfoFlow: StateFlow<ContestInfo>,
+        runsFlow: Flow<RunInfo>,
+        scoreboardFlow: Flow<Scoreboard>,
+    ) {
+        logger.info("Analytics generator service is started")
+        val runs = mutableMapOf<Int, RunAnalyse>()
+        combine(contestInfoFlow, runsFlow, scoreboardFlow, ::Triple).collect { (contestInfo, run, scoreboard) ->
```

Comment:
I'm not sure on whats happening on update, because you will receive run and scoreboard update in a random order.